Automattic/wp-calypso
# Checkout: Show helpful error message for renewal URLs with incorrect site

## Proposed Changes

When loading a checkout renewal URL where the site does not belong to the logged-in user, currently the user is redirected to their own `/me/purchases` page without an explanation. This is confusing and can cause problems if the user then tries to renew something they did not intend to.

In this PR we modify checkout so that renewal URLs (that are not gifts or siteless checkouts) with an invalid site (or with no site) will instead reach checkout and display a specific error message.

In addition, this alters checkout so that if it attempts to add renewal products from the URL and no valid renewal products exist (as is the case with an invalid site URL) it will clear the cart. This is because while the error message is being displayed, checkout will be showing the current cart for the `'no-site'` cart key, which could potentially have something in it from some previous operation. If that's the case, we don't want the user to see something in the cart that could imply their renewal URL was successful.

Fixes https://github.com/Automattic/wp-calypso/issues/75240

<img width="895" alt="Screenshot 2023-04-26 at 4 46 36 PM" src="https://user-images.githubusercontent.com/2036909/234698417-b431b2cf-d537-43b8-983b-985cab1bd406.png">

## Testing Instructions

Automated tests are included but you can test this manually.

- Visit an invalid renewal URL like `/checkout/personal-bundle/renew/12345/example.com`.
- Verify that you see an error message that tells you that the renewal is invalid and that the cart is empty.

- Add a gift purchase to your cart (instructions for gift purchases can be found in D92011-code). The URL will look something like `/checkout/personal-bundle/gift/12345`.
- Verify that checkout loads without an error message and that you see the "Gift" badge next to the product in the cart.
